### DISCUSSION TITLE
Code splitting between base ORM definitions and business logic

### DISCUSSION THREAD

**Discussion by gjeusel** [sqlalchemy/sqlalchemy#9263] [Usage Questions]
Hey there,

#### Question

I'm searching some advices on how to organise a large codebase.
Specifically, how to split business logic from the model definitions in a way that permits the usage of mixins with methods for which we don't have to import dynamically the models it uses.

#### More context

Let me provide you some (dumb) code to illustrate what is currently being done.

- `coremodels.py`:
```python
from __future__ import annotations

import sqlalchemy as sa
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship

from .business_logic import get_books_borrowed_for_author


class Base(DeclarativeBase):
    pass


class Reader(Base):
    __tablename__ = "reader"
    id = mapped_column(sa.Integer, primary_key=True)

    get_books_borrowed_for_author = get_books_borrowed_for_author


class Book(Base):
    __tablename__ = "book"
    id = mapped_column(sa.Integer, primary_key=True)
    author: Mapped[str]

    borrowed_by_id: Mapped[int] = mapped_column(sa.ForeignKey("reader.id"))
    borrowed_by: Mapped["Reader"] = relationship(back_populates="books")
```

- `business_logic.py`
```python
from __future__ import annotations

import sqlalchemy as sa
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .models import Reader, Book


def get_books_borrowed_for_author(
    reader: Reader, author: str, session: sa.orm.Session
) -> list[Book]:
    """Dumb method, I know, but it's only to make my point !"""
    from .coremodels import Book

    qry_books = sa.select(Book).where(Book.borrowed_by_id == reader.id)
    books = session.execute(qry_books).all()
    return books
```

#### Ideas

Is there any trick to maybe re-define the class associated to existing mappers like so:
- `models.py`
```python
from .coremodels import Reader as BaseReader
from .business_logic import get_books_borrowed_for_author


class Reader(BaseReader):
    get_books_borrowed_for_author = get_books_borrowed_for_author
```

?

**Comment by zzzeek** [comment]
I would think you can apply mixins in a pretty straightforward way using patterns documented at https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html , the method in question is not even sqlalchemy-specific, so just using regular mixins seems fine

business logic:

```
class HasBooksForAuthorMixin:
    def get_books_borrowed_for_author(self, ...):
```

models:

```
class Reader(HasBooksForAuthorMixin, BaseReader):
   ...
```


seems straightforward

**Reply by gjeusel** [reply]
I apologies for not being quite clear on my first draw of the question !

---

### Context

The method in question is not even sqlalchemy-specific indeed.
Although it does make use of already defined models inside it with the `sa.select(Book)`.

But my wish is to be able to define those mixins on the side, living in different modules, so I can separate the core models definitions, with the logic available on each of those models.

I'm thinking about having all my model classes defined twice:
- Once to express the tablename, columns and relationships. Those classes would then be imported inside my `business_logic.py` (statically) and be used inside the methods of the mixins.
- And a second time, inheriting from the base ones and the mixins. They will then be used everywhere else in the codebase (API layer / CLI / ...). Ideally, they would "takeover" the previously defined ones at query time. Meaning even inside `business_logic.py`, the returned instances would have those mixin methods defined on them.

Note: I can't find my way around using declarative mixins (with `__abstract__ = True` and `declared_attr`).
1) I find difficulties defining the relationships on those abstract types
2) They can't be used inside `sa.select`

---

### Code Snippet

Let me provide you a code snippet (for simplicity inside a single file):

<details>
<summary>Runnable one file snippet</summary>

```python
from __future__ import annotations

import sqlalchemy as sa
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship

metadata = sa.schema.MetaData()
registry = sa.orm.registry(metadata=metadata)


class Base(DeclarativeBase):
    registry = registry


class Reader(Base):
    __tablename__ = "reader"
    id = mapped_column(sa.Integer, primary_key=True)


class Book(Base):
    __tablename__ = "book"
    id = mapped_column(sa.Integer, primary_key=True)
    author: Mapped[str]

    borrowed_by_id: Mapped[int] = mapped_column(sa.ForeignKey("reader.id"))
    borrowed_by: Mapped["Reader"] = relationship(back_populates="books")


class HasBooksForAuthorMixin:
    def get_books_borrowed_for_author(
        self, author: str, session: sa.orm.Session
    ) -> list[Book]:
        qry_books = sa.select(Book).where(Book.borrowed_by_id == self.id)
        books = session.execute(qry_books).all()
        return books


class Reader(HasBooksForAuthorMixin, Reader):
    pass


registry.configure()


def seed(session: sa.orm.Session) -> None:
    mike = Reader()
    book = Book(author="MrSmith", borrowed_by=mike)
    session.add_all((book, mike))
    session.commit()


if __name__ == "__main__":
    engine = sa.create_engine(url="postgresql+psycopg://localhost:5432/orm-testing")

    # -- Setup + Seed
    with sa.orm.Session(bind=engine) as session:
        assert session.bind  # for type checking
        metadata.drop_all(session.bind)
        metadata.create_all(session.bind)

        seed(session)

    # -- Query
    with sa.orm.Session(bind=engine) as session:
        mike = session.execute(sa.select(Reader)).scalar()
        books = mike.get_books_borrowed_for_author(author="MrSmith", session=session)
        assert len(books) == 1
```

</details>

In this state, when running it, the following happen:
- a warning (does not fail, but worth noting)
```
SAWarning: This declarative base already contains a class with the same class name and module name as __main__.Reader, and will be replaced in the string-lookup table.
  class Reader(HasBooksForAuthorMixin, Reader)
 ```
 
 - an exception is raised at mapper configuration time:
 ```
 sqlalchemy.exc.InvalidRequestError: Multiple classes found for path "Reader" in the registry of this declarative base. Please use a fully module-qualified path.
 ```
 
 If I were to name the 2 classes differently (say `BaseReader(Base)` and `Reader(HasBooksForAuthorMixin, BaseReader)`), I would have the following error:
 ```
 sqlalchemy.orm.exc.FlushError: Attempting to flush an item of type <class '__main__.Reader'> as a member of collection "Book.borrowed_by". Expected an object of type <class '__main__.BaseReader'> or a polymorphic subclass of this type. If <class '__main__.Reader'> is a subclass of <class '__main__.BaseReader'>, configure mapper "Mapper[BaseReader(reader)]" to load this subtype polymorphically, or set enable_typechecks=False to allow any subtype to be accepted for flush.
```

### Re formulated question

At configuration time, the mappers are as follow:
```python
pdb> registry.mappers
frozenset({<Mapper at 0x10f34f490; Book>, <Mapper at 0x10f34f3a0; Reader>, <Mapper at 0x10f34e530; Reader>})
```

Is there some dark magic I can use to alter the registry and *replace* the first `<Mapper ... Reader>` by the second ?
So the first class `Reader` and the second one link to the same Mapper ?

**Reply by zzzeek** [reply]
this is not dark magic at all, you only need be more explicit which class is to be actually used to interact with the database table for this particular hierarchy.   As the base Reader intends to be subclassed, where I'd assume things like the ``books`` relationship would be added, set it up using [`__abstract__`](https://docs.sqlalchemy.org/en/20/orm/declarative_config.html#abstract):

```py

# this would ideally be called AbstractReader
class Reader(Base):
    __tablename__ = "reader"
    __abstract__ = True
    id = mapped_column(sa.Integer, primary_key=True)


class Reader(HasBooksForAuthorMixin, Reader):
    books: Mapped["Book"] = relationship(back_populates="borrowed_by")

```

The other way is to have Reader mapped at both levels, which you can just do here also, they just have to either a. have different class names or b. be in different module source files.   When you want to refer to classes in a relationship using a specific module path, use a dot-separated name as shown at https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#late-evaluation-of-relationship-arguments .

For example your test runs if I do this:

```py
class BaseReader(Base):
    __tablename__ = "reader"
    id = mapped_column(sa.Integer, primary_key=True)




class Book(Base):
    __tablename__ = "book"
    id = mapped_column(sa.Integer, primary_key=True)
    author: Mapped[str]

    borrowed_by_id: Mapped[int] = mapped_column(sa.ForeignKey("reader.id"))
    borrowed_by: Mapped["Reader"] = relationship(back_populates="books")


class HasBooksForAuthorMixin:
    def get_books_borrowed_for_author(
        self, author: str, session: sa.orm.Session
    ) -> list[Book]:
        qry_books = sa.select(Book).where(Book.borrowed_by_id == self.id)
        books = session.execute(qry_books).all()
        return books


class Reader(HasBooksForAuthorMixin, BaseReader):
    books: Mapped["Book"] = relationship(back_populates="borrowed_by")

```

that's fine as well.

What cannot happen without awkwardness is that you want Book.borrowed_by to point to BaseReader, but then magically load Reader objects from it in some other context.

**Reply by gjeusel** [reply]
Going down this road, I can't use the AbstractBook inside `get_books_borrowed_for_author` to query on:

```python
class AbstractReader(Base):
    __abstract__ = True
    __tablename__ = "reader"
    id = mapped_column(sa.Integer, primary_key=True)


class AbstractBook(Base):
    __abstract__ = True
    __tablename__ = "book"
    id = mapped_column(sa.Integer, primary_key=True)
    author: Mapped[str] = mapped_column(sa.String)

    borrowed_by_id: Mapped[int] = mapped_column(sa.ForeignKey("reader.id"))


class HasBooksForAuthorMixin:
    def get_books_borrowed_for_author(
        self, author: str, session: sa.orm.Session
    ) -> list[Book]:
        qry_books = sa.select(AbstractBook).where(AbstractBook.borrowed_by_id == self.id)
        books = session.execute(qry_books).all()
        return books


class Book(AbstractBook):
    borrowed_by: Mapped["Reader"] = relationship("Reader")


class Reader(HasBooksForAuthorMixin, AbstractReader):
    books: Mapped["Book"] = relationship("Book")
```

```
sqlalchemy.exc.ArgumentError: Column expression, FROM clause, or other columns clause element expected, got <class '__main__.AbstractBook'>.
```

My goal is to avoid having to dynamically import `Book` inside `get_books_borrowed_for_author` being defined before all the actual classes, inside another file.
(Assuming all classes will have mixins potentially using any ORM class to `sa.select` in some methods.)

Example file structure:
- **myproject.models.base**:                `Reader(Base)` + `Book(Base)` definitions
- **myproject.models.logic.reader**:    `ReaderMixin` with methods using `base.Book`
- **myproject.models.logic.book**:       `BookMixin` with methods using `base.Reader`
- **myproject.models.final**:                 `Reader(ReaderMixin, base.Reader)` + `Book(BookMixin, base.Book)`
- **myproject.models**:                          exposing models in `myproject.models.final`

**Reply by zzzeek** [reply]
you would need to acquire the correct target class dynamically, here is one way

```py
class HasBooksForAuthorMixin:
    def get_books_borrowed_for_author(
        self, author: str, session: sa.orm.Session
    ) -> list[Book]:
        Book = sa.inspect(self).mapper.attrs.books.mapper.class_
        qry_books = sa.select(Book).where(Book.borrowed_by_id == self.id)
        books = session.execute(qry_books).all()
        return books


class Book(AbstractBook):
    borrowed_by: Mapped["Reader"] = relationship("Reader", back_populates="books")


class Reader(HasBooksForAuthorMixin, AbstractReader):
    books: Mapped["Book"] = relationship("Book", back_populates="borrowed_by")

```

**Reply by gjeusel** [reply]
I'm a bit saddened by this solution as I would then loose the type checking !
In any case, thank you very much for your time and reactivity !

**Reply by zzzeek** [reply]
where is type checking lost exactly ?   inside the method you can cast() it if that's where you are referring

**Reply by gjeusel** [reply]
Yes indeed ! I can still cast it back inside the method
